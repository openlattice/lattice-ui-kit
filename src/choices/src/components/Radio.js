// @flow
import React from 'react';

import {
  ChoiceInnerWrapper,
  ChoiceLabel,
  ChoiceWrapper,
  RadioIndicator,
  RadioInput
} from './styled';

type Props = {
  checked ? :boolean;
  disabled ? :boolean;
  id ? :string;
  label ? :string;
  mode ? :string;
  name ? :string;
  onBlur ? :(event :SyntheticInputEvent<HTMLInputElement>) => void;
  onChange ? :(event :SyntheticInputEvent<HTMLInputElement>) => void;
  onFocus ? :(event :SyntheticInputEvent<HTMLInputElement>) => void;
  readOnly ? :boolean;
  value ? :any;
};

/* eslint-disable react/jsx-props-no-spreading */
const Radio = ({
  disabled,
  id,
  label,
  mode,
  readOnly,
  ...rest
} :Props) => (
  <ChoiceLabel htmlFor={id} readOnly={readOnly} disabled={disabled}>
    <ChoiceWrapper>
      <ChoiceInnerWrapper>
        <RadioInput
            id={id}
            readOnly={readOnly}
            disabled={disabled || readOnly}
            // $FlowFixMe
            {...rest} />
        <RadioIndicator mode={mode}>
          {
            mode === 'button' && label
          }
        </RadioIndicator>
      </ChoiceInnerWrapper>
    </ChoiceWrapper>
    {
      mode !== 'button' && label
    }
  </ChoiceLabel>
);
/* eslint-enable */

Radio.defaultProps = {
  checked: undefined,
  disabled: false,
  id: undefined,
  label: undefined,
  mode: undefined,
  name: undefined,
  onBlur: undefined,
  onChange: undefined,
  onFocus: undefined,
  readOnly: undefined,
  value: undefined,
};

export default Radio;
