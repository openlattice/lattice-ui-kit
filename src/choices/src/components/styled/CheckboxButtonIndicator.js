import styled from 'styled-components';

import choiceButtonStyles from './ChoiceButtonStyles';

import { duration } from '../../../../style/transitions';

const CheckboxButtonIndicator = styled.span`
  transition: background-color ${duration.swift} ease-out,
    border-color ${duration.swift} ease-out,
    box-shadow ${duration.swift} ease-out,
    color ${duration.swift} ease-out;

  ${choiceButtonStyles};
`;

export default CheckboxButtonIndicator;
