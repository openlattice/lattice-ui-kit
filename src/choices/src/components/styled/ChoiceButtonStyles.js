import { css } from 'styled-components';

import {
  defaultActive,
  defaultDisabled,
  defaultHover,
  defaultStyle,
} from '../../../../button/src/components/styled/DefaultButtonStyles';
import { PURPLE } from '../../../../colors';

const choiceButtonStyles = css`
  ${defaultStyle}
  align-items: center;
  border-radius: 3px;
  display: flex;
  flex-direction: column;
  font-weight: 600;
  justify-content: center;
  min-height: 40px;
  min-width: 84px;
  padding: 10px 12px;
  text-align: center;
  width: 100%;

  input:focus ~ &,
  input:hover ~ & {
    ${defaultHover}
  }

  input:active ~ & {
    ${defaultActive};
  }

  input:checked ~ &,
  input[readonly]:checked:disabled ~ &,
  input[readonly]:checked:hover ~ & {
    background-color: ${PURPLE.P00};
    color: ${PURPLE.P300};
  }

  input:checked:focus ~ &,
  input:checked:hover ~ & {
    background-color: ${PURPLE.P100};
  }

  input:checked:active ~ & {
    background-color: ${PURPLE.P200};
    color: ${PURPLE.P400};
  }

  input:disabled ~ & {
    ${defaultDisabled}
    cursor: not-allowed;
  }

  input:checked:disabled ~ & {
    ${defaultActive}
    cursor: not-allowed;
  }

  input[readonly] ~ & {
    cursor: default;
    pointer-events: none;
  }
`;

export default choiceButtonStyles;
