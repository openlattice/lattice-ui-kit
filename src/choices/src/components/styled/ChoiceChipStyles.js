import { css } from 'styled-components';

import { NEUTRAL, PURPLE } from '../../../../colors';

// Not really sure why, but 1px box-shadow looks significantly larger than 1px on chrome
// Using fractional pixel seems to look closer to intended design

const ChoiceChipStyles = css`
  align-items: center;
  background-color: ${NEUTRAL.N50};
  border: 3px solid transparent;
  background-clip: padding-box;
  border-radius: 16px;
  color: ${NEUTRAL.N900};
  display: flex;
  flex-direction: column;
  height: 32px;
  justify-content: center;
  padding: 10px 12px;
  text-align: center;
  user-select: none;

  input:focus-visible ~ &,
  input:hover ~ & {
    background-color: ${NEUTRAL.N100};
    cursor: pointer;
  }

  input:checked ~ &,
  input[readonly]:checked:disabled ~ &,
  input[readonly]:checked:hover ~ & {
    background-color: ${PURPLE.P00};
    color: ${PURPLE.P300};
  }

  input:focus-visible ~ & {
    box-shadow: ${NEUTRAL.N400} 0 0 0 0.1px;
  }
  input:checked:focus-visible ~ & {
    box-shadow: ${PURPLE.P300} 0 0 0 0.1px;
  }

  input:disabled ~ & {
    background-color: ${NEUTRAL.N00};
    color: ${NEUTRAL.N500};
    cursor: not-allowed;
  }

  input:checked:disabled ~ & {
    background-color: ${NEUTRAL.N400};
    color: ${NEUTRAL.N900};
    cursor: not-allowed;
  }

  input[readonly] ~ & {
    cursor: default;
    pointer-events: none;
  }
`;

export default ChoiceChipStyles;
