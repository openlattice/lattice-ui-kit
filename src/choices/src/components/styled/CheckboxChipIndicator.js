import styled from 'styled-components';

import choiceChipStyles from './ChoiceChipStyles';

import { duration } from '../../../../style/transitions';

const CheckboxChipIndicator = styled.span`
  transition: background-color ${duration.swift} ease-out,
    border-color ${duration.swift} ease-out,
    box-shadow ${duration.swift} ease-out,
    color ${duration.swift} ease-out;

  ${choiceChipStyles};
`;

export default CheckboxChipIndicator;
