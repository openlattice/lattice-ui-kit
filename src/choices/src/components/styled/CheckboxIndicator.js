import styled, { css } from 'styled-components';
import { PURPLES, WHITE } from '../../../../colors';
import choiceIndicatorStyles from './ChoiceIndicatorStyles';

const afterStyles = css`
  content: '';
  border: solid ${WHITE};
  top: 4px;
  width: 5px;
  border-width: 0 2px 2px 0;
  box-sizing: border-box;
  height: 10px;
  left: 8px;
  position: absolute;
  transform: rotate(45deg) scale(1);
  transition: transform 0.2s ease-in-out, opacity 0.2s ease-in-out;
`;

const CheckboxIndicator = styled.div`
  background-color: ${PURPLES[6]};
  border-radius: 3px;
  height: 20px;
  left: 0;
  position: absolute;
  top: 0;
  transition: background-color 0.1s ease-in-out, box-shadow 0.1s ease-in-out;
  width: 20px;

  ${choiceIndicatorStyles};

  input:not(:checked) ~ &:after,
  input:checked ~ &:after {
    ${afterStyles}
  }

  input:not(:checked) ~ &:after {
    opacity: 0;
    transform: rotate(45deg) scale(0);
  }

  input:checked ~ &:after {
    opacity: 1;
    transform: rotate(45deg) scale(1);
  }
`;

export default CheckboxIndicator;
