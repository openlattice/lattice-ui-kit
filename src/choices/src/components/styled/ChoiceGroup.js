import styled from 'styled-components';

const ChoiceGroup = styled.div`
  display: grid;
  grid-gap: 10px;
  grid-template-columns: repeat(auto-fit, minmax(120px, 1fr));
  flex-wrap: wrap;
`;

export default ChoiceGroup;
