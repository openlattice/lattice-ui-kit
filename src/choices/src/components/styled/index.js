// @ flow
import CheckboxIndicator from './CheckboxIndicator';
import ChoiceLabel from './ChoiceLabel';
import RadioIndicator from './RadioIndicator';
import { CheckboxInput, RadioInput } from './ChoiceInputs';

export {
  CheckboxIndicator,
  CheckboxInput,
  ChoiceLabel,
  RadioIndicator,
  RadioInput
};
