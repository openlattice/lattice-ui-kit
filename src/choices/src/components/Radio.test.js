import React from 'react';
import toJson from 'enzyme-to-json';
import { mount } from 'enzyme';

import Radio from './Radio';
import { RadioInput, RadioIndicator, ChoiceLabel } from './styled';

describe('Radio', () => {

  test('render matches snapshot', () => {
    const wrapper = mount(<Radio />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should pass through any props to the input', () => {
    const nativeProps = {
      disabled: true,
      name: 'test',
      required: true,
      form: 'test-form',
    };

    // eslint-disable-next-line react/jsx-props-no-spreading
    const wrapper = mount(<Radio {...nativeProps} />)
      .find('input')
      .props();
    expect(wrapper).toEqual(expect.objectContaining(nativeProps));
  });

  describe('Radio', () => {

    test('should be unchecked by default', () => {
      const wrapper = mount(<Radio />);
      expect(wrapper.find('input').prop('checked')).toEqual(undefined);
    });

    test('should render ChoiceLabel', () => {
      const wrapper = mount(<Radio label="test" />);
      expect(wrapper.find(ChoiceLabel).text()).toBe('test');
    });

    test('should render RadioIndicator', () => {
      const wrapper = mount(<Radio />);
      expect(wrapper.find(RadioIndicator).exists()).toBe(true);
    });

    test('should render RadioInput', () => {
      const wrapper = mount(<Radio />);
      expect(wrapper.find(RadioInput).exists()).toBe(true);
    });

  });

  describe('onChange', () => {

    test('should call onChange once', () => {
      const mockOnChange = jest.fn();
      const wrapper = mount(<Radio onChange={mockOnChange} />);

      wrapper.find(RadioInput).simulate('change', { target: { checked: true } });
      expect(mockOnChange).toHaveBeenCalledTimes(1);
    });

  });

});
