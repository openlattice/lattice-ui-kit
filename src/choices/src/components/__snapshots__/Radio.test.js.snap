// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio render matches snapshot 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #555e6f;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin: 10px 0;
  min-height: 20px;
  padding-left: 30px;
  pointer-events: auto;
  position: relative;
}

.c2 {
  background-color: #e6e6f7;
  border-radius: 100%;
  height: 20px;
  left: 0;
  position: absolute;
  top: 0;
  -webkit-transition: background-color 0.1s ease-in-out,box-shadow 0.1s ease-in-out;
  transition: background-color 0.1s ease-in-out,box-shadow 0.1s ease-in-out;
  width: 20px;
}

input:hover ~ .c2 {
  background-color: #8e929b;
}

input:focus-visible ~ .c2 {
  box-shadow: #410ab5 0 0 0 2px;
}

input:checked ~ .c2 {
  background-color: #8045ff;
}

input:checked:hover ~ .c2 {
  background-color: #6124e2;
}

input:disabled ~ .c2 {
  background-color: #e6e6f7;
  cursor: not-allowed;
}

input:checked:disabled ~ .c2 {
  background-color: #b6bbc7;
}

input:not(:checked) ~ .c2:after,
input:checked ~ .c2:after {
  content: '';
  background-color: #ffffff;
  border-radius: 50%;
  height: 8px;
  left: 6px;
  position: absolute;
  top: 6px;
  width: 8px;
  -webkit-transition: -webkit-transform 0.2s ease-in-out,opacity 0.2s ease-in-out;
  -webkit-transition: transform 0.2s ease-in-out,opacity 0.2s ease-in-out;
  transition: transform 0.2s ease-in-out,opacity 0.2s ease-in-out;
}

input:not(:checked) ~ .c2:after {
  opacity: 0;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
}

input:checked ~ .c2:after {
  opacity: 1;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c1 {
  position: absolute;
  z-index: -1;
  opacity: 0;
  vertical-align: middle;
}

<Radio
  disabled={false}
>
  <ChoiceLabel>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "ChoiceLabel-sc-11kob9e-0",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "align-items:center;color:",
              "#555e6f",
              ";display:inline-flex;margin:10px 0;min-height:20px;padding-left:30px;pointer-events:",
              [Function],
              ";position:relative;",
            ],
          },
          "displayName": "ChoiceLabel",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "ChoiceLabel-sc-11kob9e-0",
          "target": "label",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <label
        className="c0"
      >
        <ChoiceInputs__RadioInput
          disabled={false}
        >
          <StyledComponent
            disabled={false}
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [
                  [Function],
                ],
                "componentStyle": ComponentStyle {
                  "componentId": "ChoiceInputs__RadioInput-sc-1fkg6yx-0",
                  "isStatic": false,
                  "lastClassName": "c1",
                  "rules": Array [
                    "position:absolute;z-index:-1;opacity:0;vertical-align:middle;",
                  ],
                },
                "displayName": "ChoiceInputs__RadioInput",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "ChoiceInputs__RadioInput-sc-1fkg6yx-0",
                "target": "input",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <input
              className="c1"
              disabled={false}
              type="radio"
            />
          </StyledComponent>
        </ChoiceInputs__RadioInput>
        <RadioIndicator>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "RadioIndicator-sc-1vlnpb4-0",
                  "isStatic": true,
                  "lastClassName": "c2",
                  "rules": Array [
                    "background-color:",
                    "#e6e6f7",
                    ";border-radius:100%;height:20px;left:0;position:absolute;top:0;transition:background-color 0.1s ease-in-out,box-shadow 0.1s ease-in-out;width:20px;",
                    "input:hover ~ &{background-color:",
                    "#8e929b",
                    ";}input:focus-visible ~ &{box-shadow:",
                    "#410ab5",
                    " 0 0 0 2px;}input:checked ~ &{background-color:",
                    "#8045ff",
                    ";}input:checked:hover ~ &{background-color:",
                    "#6124e2",
                    ";}input:disabled ~ &{background-color:",
                    "#e6e6f7",
                    ";cursor:not-allowed;}input:checked:disabled ~ &{background-color:",
                    "#b6bbc7",
                    ";}",
                    ";input:not(:checked) ~ &:after,input:checked ~ &:after{",
                    "content:'';background-color:",
                    "#ffffff",
                    ";border-radius:50%;height:8px;left:6px;position:absolute;top:6px;width:8px;transition:transform 0.2s ease-in-out,opacity 0.2s ease-in-out;",
                    "}input:not(:checked) ~ &:after{opacity:0;transform:scale(0);}input:checked ~ &:after{opacity:1;transform:scale(1);}",
                  ],
                },
                "displayName": "RadioIndicator",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "RadioIndicator-sc-1vlnpb4-0",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <div
              className="c2"
            />
          </StyledComponent>
        </RadioIndicator>
      </label>
    </StyledComponent>
  </ChoiceLabel>
</Radio>
`;
